test/test_common/environment.cc

```diff
@@ -42,7 +42,14 @@ namespace {
 
 std::string makeTempDir(std::string basename_template) {
 #ifdef WIN32
-  std::string name_template = TestEnvironment::temporaryDirectory() + "/" + basename_template;
+  std::string name_template{};
+  if (std::getenv("TMP")) {
+    name_template = TestEnvironment::getCheckedEnvVar("TMP") + "\\" + basename_template;
+  } else if (std::getenv("WINDIR")) {
+    name_template = TestEnvironment::getCheckedEnvVar("WINDIR") + "\\TEMP\\" + basename_template;
```

Comment:
Is that possible put those logic inside `TestEnvironment::temporaryDirectory()`? since there is a lot of place using `TestEnvironment::temporaryDirectory()` directly, this seems to make a different behavior and is a little duplicated here.

Major usage of `maketempDir` is at 
https://github.com/envoyproxy/envoy/pull/20027/files#diff-d57e0d3c595c92209ba7143ea867a89823f8eb2a319a6feff6e2dd752250c085L67

and

https://github.com/envoyproxy/envoy/pull/20027/files#diff-d57e0d3c595c92209ba7143ea867a89823f8eb2a319a6feff6e2dd752250c085L77

So we can invoke it as `makeTempDir(TestEnvironment::temporaryDirectory() + "/" + "base name....")`.

Or we just get rid of `makeTempDir`, since it is only about connection the directory and the name.

Then we will have consistent behavior all over the places with `TestEnvironment::temporaryDirectory()`

Reply:
`TestEnvironment::temporaryDirectory()` is basically a macro invoking `getTemporaryDirectory()` which primarily relies on the system variables Envoy suggests setting in its documentation. I don't want to change the current flow and/or order of checking the system vars.  
`maketempDir()` is the function of last resort, when the required environment variables were not defined by the user; therefore, it needs to contain some hard-coded paths in order to break the cycle in this case. Also, Linux people might rely on the `/tmp` path so I don't want to change that.

Again, although the function I modified looks similar to the others now that is also checks system environment vars, it is semantically different: it does not rely on what Envoy tells users to define as system vars, and only executes fallback logic failing all else, relying on idiosyncrasies of Windows.

Reply:
> `TestEnvironment::temporaryDirectory()` is basically a macro invoking `getTemporaryDirectory()` which primarily relies on the system variables Envoy suggests setting in its documentation. I don't want to change the current flow and/or order of checking the system vars. `maketempDir()` is the function of last resort, when the required environment variables were not defined by the user; therefore, it needs to contain some hard-coded paths in order to break the cycle in this case. Also, Linux people might rely on the `/tmp` path so I don't want to change that.
> 
> > So we can invoke it as makeTempDir(TestEnvironment::temporaryDirectory() + "/" + "base name....").
> 
> Any combination of `makeTempDir` and temporaryDirectory() invoking each other will not break the cycle.
> 
> > Or we just get rid of makeTempDir, since it is only about connection the directory and the name.
> 
> Getting rid of `makeTempDir` would mean code unnecessary duplication in two call sites.
> 

Sorry , I mean ` temporaryDirectory()` is only about return the correct directly. `makeTempDir` is just about creating a directory. Then I hope the we separated two methods.

> Again, although the function I modified looks similar to the others now that is also checks system environment vars, it is semantically different: it does not rely on what Envoy tells users to define as system vars, and only executes fallback logic failing all else, relying on idiosyncrasies of Windows.
> 

I mean in `temporaryDirectory()` you still have `WIN32` macros for different platform. Then it is just single place to define those temp directly. For example, you only defined `WINDIR`, there is a lot of tests using `temporaryDirectory()` directly and those tests is going to use another directory. That will confuse also.

Reply:
Sorry for I didn't describe clearly again. I totally understand your goal here. Just want to see if we can have consistent `temp directory`  choice for the windows platform. Here is code I'm thinking:

```
std::string makeTempDir(std::string basename_template) {
#ifdef WIN32
  char* dirname = ::_mktemp(&basename_template[0]);
  RELEASE_ASSERT(dirname != nullptr, fmt::format("failed to create tempdir from template: {} {}",
                                                 basename_template, errorDetails(errno)));
  TestEnvironment::createPath(dirname);
#else
  char* dirname = ::mkdtemp(&basename_template[0]);
  RELEASE_ASSERT(dirname != nullptr, fmt::format("failed to create tempdir from template: {} {}",
                                                 basename_template, errorDetails(errno)));
#endif
  return std::string(dirname);
}

std::string getOrCreateUnixDomainSocketDirectory() {
  const char* path = std::getenv("TEST_UDSDIR");
  if (path != nullptr) {
    return std::string(path);
  }
  // Generate temporary path for Unix Domain Sockets only. This is a workaround
  // for the sun_path limit on `sockaddr_un`, since TEST_TMPDIR as generated by
  // Bazel may be too long.
#ifdef WIN32
  return makeTempDir(TestEnvironment::temporaryDirectory() + "/" + "envoy_test_uds.XXXXXX");
#else
  return makeTempDir("/tmp/envoy_test_uds.XXXXXX");
#endif
}

std::string getTemporaryDirectory() {
  std::string temp_dir;
  if (std::getenv("TEST_TMPDIR")) {
    temp_dir = TestEnvironment::getCheckedEnvVar("TEST_TMPDIR");
  } else if (std::getenv("TMPDIR")) {
    temp_dir = TestEnvironment::getCheckedEnvVar("TMPDIR");
  }
#ifdef WIN32
  else if (std::getenv("TMP")) {
    temp_dir = TestEnvironment::getCheckedEnvVar("TMP");
  } else if (std::getenv("WINDIR")) {
    temp_dir = TestEnvironment::getCheckedEnvVar("WINDIR");
  }
#endif
  else {
    return makeTempDir("envoy_test_tmp.XXXXXX");
  }
  TestEnvironment::createPath(temp_dir);
  return temp_dir;
}
```

Then you can have consistent temp dir choose for all the tests (all the tests can consider the WINDIR env now). Since a lot of calling `TestEnvironment::temporaryDirectory()` directly., like this:
https://github.com/envoyproxy/envoy/blob/3ce38d696c2f4b080b14a40512ab571f42ac6c85/test/common/config/subscription_factory_impl_test.cc#L215

And `TestEnvironment::temporaryDirectory()` becomes the only place to do the temp directory selection.

But I think I could be wrong, due to this
https://github.com/envoyproxy/envoy/pull/20027/commits/5f77c4a5223b7740f8c488cec3d8195caeb64d89#diff-d57e0d3c595c92209ba7143ea867a89823f8eb2a319a6feff6e2dd752250c085R71-R74

The comment said it wants to avoid `TEST_TMPDIR` env, since it may be too long.  I don't know if it is also a problem for windows platform. If yes, then what I'm suggested doesn't resolve that. I think your code is the only way we can do it here.

Reply:
Ok, I now understand what you mean! :) I think I disagree then.
> want to see if we can have a consistent temp directory selection for the windows platform

You moved the TMP/WINDIR branches one up, but the order of checking is exactly the same. Previously, they would have got checked when `getTemporaryDirectory` hit its `else` block.
> all the tests can consider the WINDIR env now

Again, this would have happened when  `getTemporaryDirectory` called `makeTempDir` - nothing changes here with your code.

If the goal of your suggestion is to separate the semantics of temp path determination and actual directory creation, you'd also need to make the call to `TestEnvironment::createPath(temp_dir);` appear in only one of them.

I also think that adding the `bool ignore_test_tmp_dir=false` parameter which will only ever be `true` for the one Unix test call is not clean.

Overall, this seems like overengineering for a relatively simple problem. I'll go ahead and resolve this comment.
Thank your for your suggestions and patience explaining your thought process to me.